components/prism-core.js

```diff
@@ -234,23 +234,28 @@ var _ = {
 	},
 
 	highlightAllUnder: function(container, async, callback) {
+		var filters = [];
+
 		var env = {
 			callback: callback,
-			selector: 'code[class*="language-"], [class*="language-"] code, code[class*="lang-"], [class*="lang-"] code'
+			selector: 'code[class*="language-"], [class*="language-"] code, code[class*="lang-"], [class*="lang-"] code',
+			filter: filters.push.bind(filters)
 		};
 
 		_.hooks.run('before-highlightall', env);
 
 		var elements = container.querySelectorAll(env.selector);
 
-		for (var i=0, element; element = elements[i++];) {
-			_.highlightElement(element, async === true, env.callback);
+		for (var i = 0, element; element = elements[i++];) {
+			if (filters.every(function (f) { return f(element); })) {
+				_.highlightElement(element, async === true, env.callback);
+			}
```

Comment:
Instead of modifying the internal logic of Prism core to enable this, I think we should take advantage of Prism's built-in hooks instead. Couple of ideas for how to implement this:

1. Modify the actual API/goal of the plugin to be about modifying the `selector` to `not()` classes we want to ignore from the query.
  * Pro: smaller, more focused plugin, no internal changes to core.
  * Con: less flexibility
2. Move the `before-highlight` to after the query, include `elements` on `env`, and hook the plugin there to modify the elements.
  * Pro: take advantage of what's there, minimal changes
  * Con: changes the behavior of this hook, as people could use it to modify `selector`, could be breaking.
3. Add a new hook after the query & modify `elements`
  * Pro: take advantage of what's there, minimal changes
  * Con: two similar hooks in one part of the code that are only marginally different

Thoughts?

Reply:
Maybe a bit of explanation what my goals are for this plugin:

Its purpose is to prevent elements from being highlighted by `highlightAll` or `highlightAllUnder`. So it's a mechanism to enable more fine grained control over which elements are being highlighted by Prism's `highlightAll*` methods.
This is plugin is supposed to have zero effect on elements passed to `highlightElement`. So even elements which are excluded by `highlightAll` and `highlightAllUnder`, will be highlighted. 

__Option 1__

I suppose you mean to use the CSS `:not()` selector? That's not possible because the `:not` is a component selector and has to be applied on a component basis. So we would have to _parse_ the selector and apply the `not`. It's also very limited as something like the inheritance Prism sometimes uses (e.g. language classes) can't be implemented.

Examples: Assume that the selector is `code[class*="language-"], [class*="language-"] code` and we want to exclude all elements which have the class `.exclude`.
The correct selector would be 

    code[class*="language-"]:not(.exclude), [class*="language-"] code:not(.exclude)

This is not trivial and requires that we parse the selector to some extend. Even worse: some operation are simply impossible.
Let's say that we want to exclude all elements have are descendants of an element with the class `no-highlight`. It's impossible to do because of CSS's greedy selector matching.

- `:not(.no-highlight) code` doesn't work
- `code(:not(.no-highlight) *)` not a valid component selector

The basic problem with the "modify the selector" approach is that CSS selectors only provide a logical OR for selectors and logical NOT and AND for components. We would have to implement the selector AND and NOT ourselves which, in the case of the selector NOT, is not possible as a CSS selector.

__Option 2__

It is [breaking](https://github.com/PrismJS/prism/blob/dab7998e5d9a8b7acbff5be05058a9829a0d084a/plugins/unescaped-markup/prism-unescaped-markup.js#L9-L12).

__Option 3__

Something like this, right?

<details>

```js
highlightAllUnder: function(container, async, callback) {
	var env = {
		callback: callback,
		selector: 'code[class*="language-"], [class*="language-"] code, code[class*="lang-"], [class*="lang-"] code'
	};

	_.hooks.run('before-highlightall', env);

	env.elements = [...container.querySelectorAll(env.selector)];

	_.hooks.run('before-highlightall-name-pending', env);

	for (var i = 0, element; element = env.elements[i++];) {
		_.highlightElement(element, async === true, env.callback);
	}
},
```

</details>

The only problem I see is that the `selector` is basically useless if plugins can just add elements as they like in the second hook. This also means that filtering elements becomes a pain because you have to keep the order of the hook listeners in mind (because other listeners can just add elements). But we can't enforce that element array after the hook has to be a subset of the element array before the hook, can we?

Simple filter functions don't have this problem. But I understand that the filter chaining I implemented comes quite close to a hook

Reply:
> So we would have to parse the selector and apply the `not`.

Yeah, at the time, I was just looking at the main selector, thinking it might not be too difficult to `split(',')` then map over & append `not` as needed. But thinking about it now, that's definitely not doable.

---

If it's breaking, then option 2 is out as well.

---

Yeah, the option 3 example is definitely what I was thinking.

>  is basically useless if plugins can just add elements as they like in the second hook.

I'd almost consider this a feature? If people want to be able to add their own elements, they'll now be able to without modifying `selector`, which _could_ be easier. I don't know that we need to do that enforcement–the responsibility is on the plugin (developer) to ensure that the elements in the array are what's desired after the hooks runs. You do have the issue of ordering, if one plugin is adding and another is removing elements, but that seems like a problem we could address if it comes up (e.g. hook priorities?)

Ultimately, my thought is just that using a hook here, rather than modifying core with the specific logic, is more flexible for people and seems more like "the Prism way" of doing things.

Reply:
> the responsibility is on the plugin (developer) to ensure that the elements in the array are what's desired after the hooks runs

That I can get behind.

I guess we can just ignore the ordering issue for now as there's only going to be one plugin that uses this hook in the near future. If someone really wants to get their elements filtered by this plugin, then they can just give their elements ids and add those ids to the selector in `before-highlightall`.

It just feels weird that you can add whatever elements you want in the new hook. 
On that note: We should expose the container.

Reply:
I just need a name for the new hook...

`before-highlightall-elements`?

Thoughts?

Reply:
> It just feels weird that you can add whatever elements you want in the new hook.

Honestly, this doesn't really bother me.

> On that note: We should expose the container.

The "container" being `document`?

Reply:
The container being whatever you give to `highlightAllUnder`.